### ISSUE TITLE
Help: Teleport to Marker.. Z-axis

### ISSUE THREAD

**Issue by frodzet** [crosire/scripthookvdotnet#168]
Greetings,

So i've been trying to create a teleport menu, and i finally figured out how to get the positions of a specific marker. Below is my code:

    // Get's the marker set on map.
    internal static Blip GetMarker()
    {
        int blipIterator = Function.Call<int>(Hash._GET_BLIP_INFO_ID_ITERATOR);
        int id = 0;
        bool markerFound = false;
        for (var i = Function.Call<int>(Hash.GET_FIRST_BLIP_INFO_ID, blipIterator);
           Function.Call<int>(Hash.DOES_BLIP_EXIST, i) != 0;
           i = Function.Call<int>(Hash.GET_NEXT_BLIP_INFO_ID, blipIterator))
        {
            if (Function.Call<int>(Hash.GET_BLIP_INFO_ID_TYPE, i) != 4)
                continue;
            id = i;
            markerFound = true;
            break;
        }
        if (markerFound)
            return new Blip(id);
        return null;
    }

Now since map coordinates don't have a third axis (axis-z) it will always just be 1. But i've seen Native Trainer has a teleport to Marker function, how can i do something similar with ScriptHookVDotNet? :-) I have tried to look over the native trainer source code, but i can't duplicate the results.

**Comment by RainerHilmer** [comment]
Try
var marker = GetMarker();
var groundPos = new Vector3(marker.Position.X, marker.Position.Y, World.GetGroundZ(marker.Position);

**Comment by frodzet** [comment]
Alright, so i tried that, what i did was then call the following function like so:

Entity e = GetEntity();
Function.Call(Hash.SET_ENTITY_COORDS_NO_OFFSET, e, targetPosition.X, targetPosition.Y, targetPosition.Z);

GetEntity(); is defined in it's own static method like so:

    public static Entity GetEntity()
    {
        Ped playerPed = Game.Player.Character;
        Entity e = playerPed;

        if (playerPed.IsInVehicle())
            e = playerPed.CurrentVehicle;

        return e;
    }

There's one problem tho, the entity is ALWAYS spawned below ground?

**Comment by RainerHilmer** [comment]
I can only speculate on this. I think you need an offset. Add a positive z-offset of about +1.0 or slightly more. I'm not quite sure about this. Is one unit equals one meter? The reason for your issue is probably the center of the ped model, which is usually at the waist.

**Comment by IncoCode** [comment]
from NativeTrainer by AB:
```c#
var groundHeight = new float[]
{
	100.0f, 150.0f, 50.0f, 0.0f, 200.0f, 250.0f, 300.0f, 350.0f, 400.0f,
	450.0f, 500.0f, 550.0f, 600.0f, 650.0f, 700.0f, 750.0f, 800.0f
};
Blip blip = World.GetActiveBlips().FirstOrDefault( b => b.Type == 4 );
if ( blip != null )
{
	player.Position = blip.Position;
	foreach ( float g in groundHeight )
	{
		float z = 0;
		player.Position = new Vector3( player.Position.X, player.Position.Y, g );
		Wait( 100 );
		if ( Function.Call<bool>( Hash.GET_GROUND_Z_FOR_3D_COORD, player.Position.X,
			player.Position.Y, g, &z ) )
		{
			player.Position = new Vector3( player.Position.X, player.Position.Y, z + 3 );
			break;
		}
	}
	UI.ShowSubtitle( "Done", 2000 );
}
else
{
	UI.ShowSubtitle( "Not found", 2000 );
}
```

**Comment by IncoCode** [comment]
btw, we should change World.GetGroundZ return values.

**Comment by frodzet** [comment]
@IncoCode Agreed, and thank you very much for the code, works like a charm, hell it works even better than i anticipated! :-)

**Comment by JohnnyCrazy** [comment]
Fixed in 6fe20aa1152cb35a6143c17e4a2a8dcb6406db9d

**Comment by JohnnyCrazy** [comment]
The GroundZ thing got fixed in 6fe20aa1152cb35a6143c17e4a2a8dcb6406db9d and I guess the question is solved :)
